src/xmipp/libraries/reconstruction_cuda/cuda_forward_art_zernike3d.cpp

```diff
@@ -10,6 +10,100 @@
 #define IS_OUTSIDE2D(ImD, i, j) \
 	((j) < STARTINGX((ImD)) || (j) > FINISHINGX((ImD)) || (i) < STARTINGY((ImD)) || (i) > FINISHINGY((ImD)))
 
+// Cuda memory helper function
+namespace {
+
+	template<typename T>
+	cudaError cudaMallocAndCopy(T **target, const T *source, size_t numberOfElements, size_t memSize = 0)
+	{
+		size_t elemSize = numberOfElements * sizeof(T);
+		memSize = memSize == 0 ? elemSize : memSize * sizeof(T);
+
+		cudaError err = cudaSuccess;
+		if ((err = cudaMalloc(target, memSize)) != cudaSuccess) {
+			*target = NULL;
+			return err;
+		}
+
+		if ((err = cudaMemcpy(*target, source, elemSize, cudaMemcpyHostToDevice)) != cudaSuccess) {
+			cudaFree(*target);
+			*target = NULL;
+		}
+
+		if (memSize > elemSize) {
+			cudaMemset((*target) + numberOfElements, 0, memSize - elemSize);
+		}
+
+		return err;
+	}
+
+	void processCudaError()
```

Comment:
Let it throw an execption with error message and catch it in `reconstruction_adapt_cuda`